### ISSUE TITLE
Tensorflow build fails with vectorization 

### ISSUE THREAD

**Issue by Nayana-ibm** [tensorflow/tensorflow#22023]
### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Ubuntu 16.04, 18.04 
- **TensorFlow installed from (source or binary)**:  source 
- **TensorFlow version (use command below)**:  1.8.0 
- **Python version**:  2.7.15rc1
- **Bazel version (if compiling from source)**:  0.12.0
- **GCC/Compiler version (if compiling from source)**: gcc-7.3.0, gcc-6.3.0
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**:  ` bazel build -c opt  --incompatible_load_argument_is_label=false --incompatible_disallow_uncalled_set_constructor=false  --config=opt  --copt="-mzvector" //tensorflow/tools/pip_package:build_pip_package`

**Description:** 
While building TensorFlow with vectorization on s390x platform, faced following issues :
**With gcc 6.3.0 , Ubuntu 16.04, z13**  :

```
INFO: From Compiling external/snappy/snappy-sinksource.cc [for host]:
cc1plus: warning: command line option '-Wno-implicit-function-declaration' is valid for C/ObjC but not for C++
Slow read: a 7184608-byte read from /root/.cache/bazel/_bazel_root/9b63c66f147d91a5644a0e81d48c020e/external/org_sqlite/sqlite3.c took 26065ms.
ERROR: /root/.cache/bazel/_bazel_root/9b63c66f147d91a5644a0e81d48c020e/external/org_sqlite/BUILD:35:1: C++ compilation of rule '@org_sqlite//:org_sqlite' failed (Exit 1)
external/org_sqlite/sqlite3.c: In function 'vdbeRecordCompareInt':
external/org_sqlite/sqlite3.c:76283:1: internal compiler error: in maybe_record_trace_start, at dwarf2cfi.c:2284
}
ˆ
0x1385bc7 maybe_record_trace_start
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2284
0x1388285 scan_trace
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2462
0x13888dd create_cfi_notes
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2616
0x13888dd execute_dwarf2_frame
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2974
0x13888dd execute
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:3454
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1795.018s, Critical Path: 24.14s
FAILED: Build did NOT complete successfully
```
**With gcc 6.3.0/7.3.0 , Ubuntu 16.04, z14**  :

```
ERROR: /tensorflow/tensorflow/core/kernels/BUILD:436:1: C++ compilation of rule '//tensorflow/core/kernels:split_lib' failed (Exit 1)
In file included from external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/Core:427:0,
                from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:14,
                from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                from ./tensorflow/core/kernels/split_lib.h:20,
                from tensorflow/core/kernels/split_lib_cpu.cc:18:
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Eigen::internal::Packet4f Eigen::internal::vec_splat_packet4f(const Packet4f&)':
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/arch/ZVector/PacketMath.h:657:11: error: request for member 'v4f' in 'splat', which is of non-class type 'Eigen::internal::Packet4f {aka __vector(4) float}'
    splat.v4f[0] = vec_splat(from.v4f[0], 0);
          ˆ˜˜
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/arch/ZVector/PacketMath.h:657:35: error: request for member 'v4f' in 'from', which is of non-class type 'const Packet4f {aka const __vector(4) float}'
    splat.v4f[0] = vec_splat(from.v4f[0], 0);
                                  ˆ˜˜
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/arch/ZVector/PacketMath.h:657:44: internal compiler error: tree check: expected class 'type', have 'exceptional' (error_mark) in s390_fn_types_compatible, at config/s390/s390-c.c:727
    splat.v4f[0] = vec_splat(from.v4f[0], 0);
                                           ˆ
0x80cc535f tree_class_check_failed(tree_node const*, tree_code_class, char const*, int, char const*)
       ../../gcc-6.3.0/gcc/tree.c:9704
```

**With gcc 7.3.0 , Ubuntu 18.04, z14**  :
```

INFO: From Compiling tensorflow/contrib/lite/util.cc:
tensorflow/contrib/lite/util.cc: In function 'TfLiteIntArray* tflite::ConvertArrayToTfLiteIntArray(int, const int*)':
tensorflow/contrib/lite/util.cc:25:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for (size_t i = 0; i < rank; i++) {
                      ˜˜ˆ˜˜˜˜˜
ERROR: /scratch/ecos0030/nayana/tensorflow/tensorflow/contrib/lite/kernels/BUILD:43:1: C++ compilation of rule '//tensorflow/contrib/lite/kernels:eig         en_support' failed (Exit 1)
In file included from external/eigen_archive/Eigen/Core:248:0,
                 from ./third_party/eigen3/Eigen/Core:1,
                 from tensorflow/contrib/lite/kernels/eigen_support.cc:17:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pround(const Packet&) [with Packet = __vector(2         ) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:471:86: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet2d pround<Packet2d>(const Packet2d& a) { return vec_round(a); }
                                                                                      ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:471:86: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pceil(const Packet&) [with Packet = __vector(2)          double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:472:86: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet2d pceil<Packet2d>(const  Packet2d& a) { return vec_ceil(a); }
                                                                                      ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:472:86: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pfloor(const Packet&) [with Packet = __vector(2         ) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:473:86: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet2d pfloor<Packet2d>(const Packet2d& a) { return vec_floor(a); }
                                                                                      ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:473:86: note: suggested alternative: '__builtin_s390_vfidb'
In file included from external/eigen_archive/Eigen/Core:427:0,
                 from ./third_party/eigen3/Eigen/Core:1,
                 from tensorflow/contrib/lite/kernels/eigen_support.cc:17:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pmadd(const Packet&, const Packet&, const Packe         t&) [with Packet = __vector(4) float]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1073:141: error: cannot convert 'const Packet4f {aka const __vector(4) float}' to '__         vector(2) double' for argument '1' to '__vector(2) double __builtin_s390_vfmadb(__vector(2) double, __vector(2) double, __vector(2) double)'
template<> EIGEN_STRONG_INLINE Packet4f pmadd<Packet4f>  (const Packet4f& a, const Packet4f& b, const Packet4f& c) { return vec_madd(a, b, c); }
                                                                                                                                             ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pxor(const Packet&, const Packet&) [with Packet          = __vector(4) float]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1078:118: error: invalid parameter combination for intrinsic '__builtin_s390_vec_xor'
template<> EIGEN_STRONG_INLINE Packet4f pxor<Packet4f>   (const Packet4f& a, const Packet4f& b) { return vec_xor(a, b); }
                                                                                                                      ˆ
In file included from external/eigen_archive/Eigen/Core:248:0,
                 from ./third_party/eigen3/Eigen/Core:1,
                 from tensorflow/contrib/lite/kernels/eigen_support.cc:17:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pround(const Packet&) [with Packet = __vector(4         ) float]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1080:87: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet4f pround<Packet4f> (const Packet4f& a) { return vec_round(a); }
                                                                                       ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1080:87: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pceil(const Packet&) [with Packet = __vector(4)          float]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1081:87: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet4f pceil<Packet4f>  (const Packet4f& a) { return vec_ceil(a); }
                                                                                       ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1081:87: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pfloor(const Packet&) [with Packet = __vector(4         ) float]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1082:87: error: '__builtin_s390_vfi' was not declared in this scope
template<> EIGEN_STRONG_INLINE Packet4f pfloor<Packet4f> (const Packet4f& a) { return vec_floor(a); }
                                                                                       ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:1082:87: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/MathFunctions.h: In function 'Packet Eigen::internal::pexp(const Packet&) [with Packet = __vector(         2) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/MathFunctions.h:101:8: error: '__builtin_s390_vfi' was not declared in this scope
   fx = vec_floor(fx);
        ˆ
external/eigen_archive/Eigen/src/Core/arch/ZVector/MathFunctions.h:101:8: note: suggested alternative: '__builtin_s390_vfidb'
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

**Comment by Nayana-ibm** [comment]
We also tried to build eigen separately , the build is successful on z14.  Trying to debug the issue with TensorFlow. 
Any inputs on what could be causing the above issues?

**Comment by jdduke** [comment]
TensorFlow Lite uses a non-traditional approach for importing and including Eigen, so it's quite possible there are some platform/flag-specific issues that it's not handling properly. See also https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/lite/kernels/internal/optimized/eigen_tensor_reduced_instantiations_oss.h

**Comment by jdduke** [comment]
I'm not familiar with the sqlite dependency, but we might consider excluding TensorFlow Lite from the default s390x platform builds.

**Comment by jdduke** [comment]
+gunan per the issues with sqlite.

**Comment by gunan** [comment]
Looks like sqlite does not build on s390x?
maybe we also need to exclude sqlite ops in s390x?
Do we now that the failing code works/supports big endian?

**Comment by Nayana-ibm** [comment]
@gunan  We are observing different issues with different GCC version with vectorization:

1. With gcc 6.3.0 , Ubuntu 16.04, z13 :   Sqlite build issue
2. With gcc 6.3.0/7.3.0 , Ubuntu 16.04, z14 :  TF build failure on Eigen /internal compiler error. 
3. With gcc 7.3.0 , Ubuntu 18.04, z14 :  TF buil dfailure on Eigen. 
We are currently investigating more on Eigen failures. (Any pointers on this would be helpful)

For sqlite , tried to comment out download sqlite code from `tensorflow/workspace.bzl` but it didn't help. Is there any option in configure through which we can disable this feature?

**Comment by gunan** [comment]
I assume that sqlite build issue is everywhere. But with newer gcc versions
there is also the eigen build issue.

For sqlite, we have to conditionally exclude sqlite targets if we are
building for s390x.
Do you have a link for that build result(log)?

**Comment by Nayana-ibm** [comment]
@gunan I have tried TensorFlow build v1.8.0 on local z13 vm with gcc 6.3.0 on Ubuntu 16.04.
Below is the log for sqlite.  Please note we have not observed sqlite issue ( or may be build didn't reach till sqlite module) with z14.  

```
INFO: From Compiling external/snappy/snappy-sinksource.cc [for host]:
cc1plus: warning: command line option '-Wno-implicit-function-declaration' is valid for C/ObjC but not for C++
Slow read: a 7184608-byte read from /root/.cache/bazel/_bazel_root/9b63c66f147d91a5644a0e81d48c020e/external/org_sqlite/sqlite3.c took 26065ms.
ERROR: /root/.cache/bazel/_bazel_root/9b63c66f147d91a5644a0e81d48c020e/external/org_sqlite/BUILD:35:1: C++ compilation of rule '@org_sqlite//:org_sqlite' failed (Exit 1)
external/org_sqlite/sqlite3.c: In function 'vdbeRecordCompareInt':
external/org_sqlite/sqlite3.c:76283:1: internal compiler error: in maybe_record_trace_start, at dwarf2cfi.c:2284
}
ˆ
0x1385bc7 maybe_record_trace_start
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2284
0x1388285 scan_trace
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2462
0x13888dd create_cfi_notes
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2616
0x13888dd execute_dwarf2_frame
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:2974
0x13888dd execute
        ../../gcc-6.3.0/gcc/dwarf2cfi.c:3454
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1795.018s, Critical Path: 24.14s
FAILED: Build did NOT complete successfully
```

**Comment by gunan** [comment]
Do we know if sqlite has big endian support, by the way?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @gunan: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by duane-ibm** [comment]
With gcc 6.3.0 , Ubuntu 16.04, z13 : using bazel 0.15.0 instead of bazel 0.12.0, Build for vectorization does not fail with the `sqlite3.c` error anymore but fails with the eigen issue in PacketMath.h similarly seen on z14.
```
ERROR: /home/test/tensorflow/tensorflow/contrib/lite/kernels/BUILD:43:1: C++ compilation of rule '//tensorflow/contrib/lite/kernels:eigen_support' failed (Exit 1)
In file included from external/eigen_archive/Eigen/Core:427:0,
                 from ./third_party/eigen3/Eigen/Core:1,
                 from tensorflow/contrib/lite/kernels/eigen_support.cc:17:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Eigen::internal::Packet4f Eigen::internal::vec_splat_packet4f(const Packet4f&)':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:657:11: error: request for member 'v4f' in 'splat', which is of non-class type 'Eigen::internal::Packet4f {aka __vector(4) float}'
     splat.v4f[0] = vec_splat(from.v4f[0], 0);
           ^~~
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:657:35: error: request for member 'v4f' in 'from', which is of non-class type 'const Packet4f {aka const __vector(4) float}'
     splat.v4f[0] = vec_splat(from.v4f[0], 0);
                                   ^~~
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:657:44: internal compiler error: tree check: expected class 'type', have 'exceptional' (error_mark) in s390_fn_types_compatible, at config/s390/s390-c.c:727
     splat.v4f[0] = vec_splat(from.v4f[0], 0);
                                            ^
0x80cc5367 tree_class_check_failed(tree_node const*, tree_code_class, char const*, int, char const*)
        ../../gcc-6.3.0/gcc/tree.c:9704
```

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @gunan: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @asimshankar, @gunan: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jdduke** [comment]
Can you sync to master and try building again? We recently updated all of our Eigen dependencies, and I'd like to see if you're hitting the same issue.

**Comment by Nayana-ibm** [comment]
@jdduke I have tried building v1.12.0 with vectorization on Ubuntu 18.04 with gcc v7.3.0, Bazel v0.15.0.

command used:
` bazel --host_jvm_args="-Xms512m" --host_jvm_args="-Xmx1024m" build --config=opt --copt="-mzvector" //tensorflow/tools/pip_package:build_pip_package `

However got an error: 

```
ERROR: /root/tensorflow/tensorflow/contrib/lite/kernels/BUILD:158:1: C++ compilation of rule '//tensorflow/contrib/lite/kernels:builtin_op_kernels' failed (Exit 1)^M
In file included from external/eigen_archive/Eigen/Core:431:0,
                from ./third_party/eigen3/Eigen/Core:1,
                from ./tensorflow/contrib/lite/kernels/internal/optimized/optimized_ops.h:28,
                from tensorflow/contrib/lite/kernels/space_to_batch_nd.cc:19:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet Eigen::internal::pmin(const Packet&, const Packet&) [with Packet = __vector(2) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:454:115: *error: '__builtin_s390_vec_min' matching variant requires z14 or higher*
template<> EIGEN_STRONG_INLINE Packet2d pmin<Packet2d>(const Packet2d& a, const Packet2d& b) { return vec_min(a, b); }
```

Now, started build on z14 vm. will share the result.

**Comment by Nayana-ibm** [comment]
Tensorflow build  with vectorization failed on z14  with an error:

 ```

ERROR: /home/dev/tensorflow/tensorflow/contrib/lite/kernels/BUILD:57:1: C++ compilation of rule '//tensorflow/contrib/lite/kernels:eigen_support' failed (Exit 1)^M
In file included from external/eigen_archive/Eigen/Core:251:0,
                 from ./tensorflow/contrib/lite/kernels/internal/optimized/eigen_tensor_reduced_instantiations_oss.h:26,
                 from ./tensorflow/contrib/lite/kernels/internal/optimized/eigen_spatial_convolutions.h:37,
                 from tensorflow/contrib/lite/kernels/eigen_support.cc:20:
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet EigenForTFLite::internal::pround(const Packet&) [with Packet = __vector(2) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:471:86: error: '__builtin_s390_vfi' was not declared in this scope
 template<> EIGEN_STRONG_INLINE Packet2d pround<Packet2d>(const Packet2d& a) { return vec_round(a); }
                                                                                      ^
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:471:86: note: suggested alternative: '__builtin_s390_vfidb'
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h: In function 'Packet EigenForTFLite::internal::pceil(const Packet&) [with Packet = __vector(2) double]':
external/eigen_archive/Eigen/src/Core/arch/ZVector/PacketMath.h:472:86: error: '__builtin_s390_vfi' was not declared in this scope
 template<> EIGEN_STRONG_INLINE Packet2d pceil<Packet2d>(const  Packet2d& a) { return vec_ceil(a); }

```

**Comment by gunan** [comment]
@jdduke looks like an eigen issue?
Now that we have unfrozen our eigen version, we can actually push a fix and update our dependency.
Could you take a look?

**Comment by jdduke** [comment]
Yeah, it's interesting that this is failing within TensorFlow Lite, but not TensorFlow (which would also ahve the Eigen dependency).  @rmlarsen do you know if there's anything we ought to be doing differently to allow for vectorized Eigen builds?

**Comment by duane-ibm** [comment]
How do we build tensorflow with vectorization on Intel?
Do we need to enable it while configuring? Or do we specify some flags? Like how we do it on z13 and higher using the `--copt="-mzvector"` flag.

**Comment by duane-ibm** [comment]
@gunan any info regarding vectorization on intel, as we are facing issues with s390x, wanted to check on intel.

**Comment by gunan** [comment]
I think in x86 land, vectorization means building with avx/avx2 and avx512 support.
I do not think that has a direct translation to s390x

**Comment by Nayana-ibm** [comment]
@gunan  I tried building TensorFlow v1.12.0 with an option --copt=-DEIGEN_ENABLE_AVX512 
Is it the correct flag to enable vectorization  for x86? 
` bazel build --copt=-DEIGEN_ENABLE_AVX512 //tensorflow/tools/pip_package:build_pip_package `

I could also build TensorFlow wheel and install but got a message as below: 
```
>>> import tensorflow as tf
>>> sess = tf.Session()
: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA

```

**Comment by Nayana-ibm** [comment]
Also checking with @benoitsteiner whats the exact command for x86 through https://github.com/tensorflow/tensorflow/issues/4775

**Comment by Nayana-ibm** [comment]
@gunan @benoitsteiner an you please confirm if using flag --copt=-DEIGEN_ENABLE_AVX512 is correct on Intel for vectoriztion?

We are still struggling with TenorFlow build failures on s390x with vetorization using flag  -march=z14 -mzvector.

**Comment by Nayana-ibm** [comment]
I have tried simple example using eigen vector functions on both the platforms : Intel and s390x.
On Intel , the code can compile with vectorization (by default -msse2 ) and generates output.

On s390x, compilation failed with an error:` __builtin_s390_vfi was not declared in this scope`.
Same issue is observed while building TensorFlow with vectorization on z14 vm.

I have used -march=z14 -mzvector flags on z14 s390x vm.
Reference: http://eigen.tuxfamily.org/index.php?ti ... torization

Could you confirm if Eigen supports vectorization for s390x?

**Comment by gunan** [comment]
`--copt=-DEIGEN_ENABLE_AVX512` sounds like the right flag to use avx512 level vectorization in eigen code. However, I am quite certain that that define cannot work on s390x. Even if that is enabled, you still need `-mavx512f` flag, and last time I checked there were either build or test failures when we enabled avx512 instructions.

The question about if eigen supports s390x vectorization, I am afraid I have no answer to that. Eigen maintainers will be able to answer that. http://eigen.tuxfamily.org/bz/ should be a batter route to get a hold of them.

**Comment by Nayana-ibm** [comment]
@gunan Thanks for your inputs . 
Yes ..the --copt=-DEIGEN_ENABLE_AVX512 flag I used for Intel only.  For s390x, I tried with -march=z14 -mzvector flags.

I am trying to reach Eigen community however they have disabled sign in to new users.  Now waiting for reply for emails that I sent to mentioned id. 

Meanwhile I will try to rebuild and investigate further on vectorization for s390x.

**Comment by Andreas-Krebbel** [comment]
@Nayana-ibm 
The build problem probably is this one. We also ran into it while trying to build tensorflow:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89952
I've attached an experimental patch for GCC which is currently being tested.

Eigen has been optimized for IBM z14 using GCC builtins. It uses the set of altivec-style builtins which introduce "vector" as a keyword. This allows defining vector data types more conveniently. That's why -mzvector is required during build.
